# Đồ án 1: Thu thập dữ liệu

|Danh sách thành viên|MSSV|Phân chia công việc|Tiến độ hoàn thành|
|---|---|---|---|
|Vũ Duy|1712396|Thu thập dữ liệu bằng API|Hoàn thành|
|Phạm Nguyễn Hải Dương|1712376|Thu thập dữ liệu bằng HTML|Hoàn thành|
|Nguyễn Văn Hậu|1712258|Thu thập dữ liệu bằng API|Hoàn thành|
|Lâm Thiên Đức|1712355|Thu thập dữ liệu bằng HTML|Hoàn thành|
|Hồ Hoàng Duy|19120207|Viết notebook, test chương trình, viết mô tả ý nghĩa dữ liệu thu thập được|Hoàn thành|

# Nhiệm vụ:
## Thu nhập các thông tin từ trang https://soundcloud.com/ bằng HTML và API

### Thu thập dữ liệu bằng API

Xem tại file notebook: Project01_api.ipynb

### Thu thập dữ liệu bằng HTML

Xem tại file notebook: Project01_Crawl.ipynb
Lưu ý: để chạy notebook, máy tính cần phải cài đặt sẵn chromedriver phù hợp với phiên bản của trình duyệt Google Chrome. Tham khảo tại: https://chromedriver.chromium.org/

# Khám phá dữ liệu

## Do thu thập dữ liệu từ API sẽ cho nhiều trường dữ liệu hơn nên nhóm sẽ mô tả thuộc tính trong tập dữ liệu thu thập được bằng API.

In [1]:
import numpy as np
import pandas as pd

## 1. Tập dữ liệu users

### Kiểm tra tập dữ liệu vừa thu thập được

In [2]:
raw_data_user = pd.read_csv('Api_data/user.csv', sep=',')

In [3]:
raw_data_user.shape

(1499, 32)

Vậy trong file `user.csv` ta có 32 thuộc tính và 1499 mẫu dữ liệu.

|Thuộc tính|Kiểu dữ liệu|Ý nghĩa|Khoảng giá trị|
|---|---|---|---|
|avatar_url|Chuỗi|Ảnh đại diện của user|---|
|city|Chuỗi|Thành phố user đang sống|---|
|comments_count|Số nguyên|Số lượt bình luận của user|Liên tục|
|country_code|Chuỗi|Mã quốc gia user đang sống|---|
|created_at|Thời gian|Thời gian khởi tạo tài khoản|---|
|creator_subscriptions|Chuỗi|Loại tài khoản của user|---|
|description|Chuỗi|Thông tin mô tả bản thân của user|---|
|followers_count|Số nguyên|Số người đang theo dõi user|Liên tục|
|followings_count|Số nguyên|Số người user đang theo dõi|Liên tục|
|first_name|Chuỗi|Tên của user|---|
|full_name|Chuỗi|Họ tên của user|---|
|groups_count|Số nguyên|Số nhóm user đang tham gia|Liên tục|
|id|Số nguyên|Mã định danh user trên hệ thống|---|
|kind|Chuỗi|Loại đối tượng|user|
|last_modified|Thời gian|Thời gian của lần chỉnh sửa gần nhất|---|
|last_name|Chuỗi|Họ của user|---|
|likes_count|Số nguyên|Số lượt thích của user|Liên tục|
|playlist_likes_count|Số nguyên|Số playlist mà user đã thích|Liên tục|
|permalink|Chuỗi|Tên đường dẫn cố định của user|---|
|permalink_url|Chuỗi|Tên đường dẫn tuyệt đối của user|---|
|playlist_count|Số nguyên|Số playlist của user|Liên tục|
|repost_count|Số nguyên|Số bài được user đăng lại|Liên tục|
|track_count|Số nguyên|Số track của user|Liên tục|
|uri|Chuỗi|Đường dẫn API của user|---|
|urn|Chuỗi|Đường dẫn API của user|---|
|username|Chuỗi|Tên đăng nhập của user|---|
|verified|Luận lý|Cho biết tài khoản của user đã được xác minh hay chưa|True hoặc False|
|visuals|Chuối|Cho biết giá trị các id: urn, enabled, visuals, tracking|---|
|station_urn|Chuỗi|---|---|
|station_permalink|Chuỗi|---|---|
|creator_subscription.product.id|Chuỗi|Phân loại sản phẩm của thành viên|---|
|badges|Chuỗi|Cho biết user có phải pro hay không, đã verified chưa|---|

Kiểm tra số lượng của một số mẫu đã thu thập được

In [4]:
print(f"Số lượng id: {raw_data_user['id'].nunique()}")
print(f"Số lượng permalink_url: {raw_data_user['permalink_url'].nunique()}")
print(f"Số lượng avatar_url: {raw_data_user['avatar_url'].nunique()}")
print(f"Số lượng first_name: {raw_data_user['first_name'].nunique()}")
print(f"Số lượng last_name: {raw_data_user['last_name'].nunique()}")
print(f"Số lượng full_name: {raw_data_user['full_name'].nunique()}")
print(f"Số lượng username: {raw_data_user['username'].nunique()}")
print(f"Số lượng permalink_url: {raw_data_user['permalink_url'].nunique()}")
print(f"Số lượng permalink: {raw_data_user['permalink'].nunique()}")
print(f"Số lượng uri: {raw_data_user['uri'].nunique()}")              
print(f"Số lượng urn: {raw_data_user['urn'].nunique()}")    

Số lượng id: 1499
Số lượng permalink_url: 1499
Số lượng avatar_url: 1205
Số lượng first_name: 861
Số lượng last_name: 1009
Số lượng full_name: 1185
Số lượng username: 1499
Số lượng permalink_url: 1499
Số lượng permalink: 1499
Số lượng uri: 1499
Số lượng urn: 1499


Trong dữ liệu thu thập được thì có nhiều thuộc tính gần như nhau về mặt ý nghĩa: `creator_subscriptions`, `badges` đều mang ý nghĩa phân loại người dùng đã đăng ký

In [5]:
raw_data_user[['creator_subscriptions', 'badges']]

creator_subscriptions  \
0     [{'product': {'id': 'creator-pro-unlimited'}}]   
1     [{'product': {'id': 'creator-pro-unlimited'}}]   
2                      [{'product': {'id': 'free'}}]   
3                      [{'product': {'id': 'free'}}]   
4                      [{'product': {'id': 'free'}}]   
...                                              ...   
1494                   [{'product': {'id': 'free'}}]   
1495                   [{'product': {'id': 'free'}}]   
1496                   [{'product': {'id': 'free'}}]   
1497                   [{'product': {'id': 'free'}}]   
1498                   [{'product': {'id': 'free'}}]   

                                                 badges  
0     {'pro': False, 'pro_unlimited': True, 'verifie...  
1     {'pro': False, 'pro_unlimited': True, 'verifie...  
2     {'pro': False, 'pro_unlimited': False, 'verifi...  
3     {'pro': False, 'pro_unlimited': False, 'verifi...  
4     {'pro': False, 'pro_unlimited': False, 'verifi...  
...                                                 ...  
1494  {'pro': False, 'pro_unlimited': False, 'verifi...  
1495  {'pro': False, 'pro_unlimited': False, 'verifi...  
1496  {'pro': False, 'pro_unlimited': False, 'verifi...  
1497  {'pro': False, 'pro_unlimited': False, 'verifi...  
1498  {'pro': False, 'pro_unlimited': False, 'verifi...  

[1499 rows x 2 columns]

In [6]:
raw_data_user['creator_subscriptions'].unique()

array(["[{'product': {'id': 'creator-pro-unlimited'}}]",
       "[{'product': {'id': 'free'}}]",
       "[{'product': {'id': 'creator-pro'}}]"], dtype=object)

Ta có thể nhận thấy thuộc tính `creator_subscriptions` đều có khóa dạng `product.id`nên ta có thể chọn luôn thuộc tính cụ thể `creator_subcription.product.id` đại diện là được.

Ta cùng xem những giá trị của các thuộc tính còn lại trong dữ liệu

In [7]:
print(f"Số lượng groups_count: {raw_data_user['groups_count'].nunique()}")
print(f"Số lượng kind: {raw_data_user['kind'].nunique()}")
print(f"Số lượng verified: {raw_data_user['verified'].nunique()}")
print(f"Số lượng visuals: {raw_data_user['visuals'].nunique()}")
print(f"Số lượng badges: {raw_data_user['badges'].nunique()}")
print(f"Số lượng reposts_count: {raw_data_user['reposts_count'].nunique()}")

Số lượng groups_count: 1
Số lượng kind: 1
Số lượng verified: 2
Số lượng visuals: 442
Số lượng badges: 6
Số lượng reposts_count: 0


## 2. Tập dữ liệu playlists

### Kiểm tra tập dữ liệu vừa thu thập được

In [8]:
raw_data_playlist = pd.read_csv('Api_data/playlist.csv', sep=',')

In [9]:
raw_data_playlist.shape

(1499, 33)

Tập dữ liệu playlists có 1499 mẫu và 33 thuộc tính.

Mô tả một số thuộc tính trong tập dữ liệu

|Thuộc tính|Kiểu dữ liệu|Ý nghĩa|Miền giá trị|
|----------|------------|-------|------------|
|artwork_url|Chuỗi|Đường dẫn đến ảnh đại diện của playlist||
|created_at|Thời gian|Thời gian tạo playlist||
|description|Chuỗi|Mô tả playlist của người dùng||
|duration|Số nguyên|Tổng thời gian của các track nhạc trong playlist|Liên tục (đơn vị: mili giây)|
|embeddable_by|Chuỗi|Đối tượng có quyền nhúng track hoặc playlist này|"all", "me" hoặc None|
|genre|Chuỗi|Thể loại nhạc của playlist||
|id|Kiểu phân loại|ID của playlist||
|kind|Chuỗi|Loại đối tượng|playlist|
|label_name|Chuỗi|Hãng thu âm||
|last_modified|Thời gian|Lần chỉnh sửa gần nhất|   |
|license|Chuỗi|Giấy phép Creative Commons|“no-rights-reserved”, “all-rights-reserved”, “cc-by”, “cc-by-nc”, “cc-by-nd”, “cc-by-sa”, “cc-by-nc-nd”, “cc-by-nc-sa”|
|likes_count|Số nguyên|Số lượt yêu thích của playlist|Liên tục|
|managed_by_feeds|Luận lý|Playlist có được quản lý bởi RSS feed|True hoặc False|
|permalink|Chuỗi|Đường dẫn cố định của tài nguyên||
|permalink_url|Chuỗi|Đường dẫn đến Soundcloud||
|public|Luận lý|Playlist có công khai hay không|True hoặc False. Trong dataset chỉ chứa playlist có giá trị True do chỉ rút trích được các playlist chia sẻ công khai|
|purchase_title|Chuỗi|||
|purchase_url|Chuỗi|||
|release_date|Thời gian|Thời gian xuất bản playlist||
|reposts_count|Số nguyên|Số lượt đăng lại playlist|Liên tục|
|secret_token|Chuỗi|||
|sharing|Chuỗi|Chế độ chia sẻ (công khai hoặc riêng tư)|public, private|
|tag_list|Chuỗi|Danh sách các tag của playlist||
|title|Chuỗi|Tiêu đề playlist||
|uri|Chuỗi|Đường dẫn API của playlist||
|user_id|Kiểu phân loại|ID người sở hữu playlist||
|set_type|Chuỗi|Định dạng đĩa|"album", "compilation", "ep", "single"|
|is_album|Luận lý|Playlist có phải album hay không|True hoặc False|
|published_at|Thời gian|Thời điểm đăng playlist||
|display_date|Thời gian|Thời điểm playlist hiển thị||
|track_counts|Số nguyên|Số lượng track trong playlist|Liên tục|
|tracks|Chuỗi|Danh sách id của các track trong playlist||

Kiểm tra số lượng mẫu

In [10]:
print(f"Số lượng id: {raw_data_playlist['id'].nunique()}")
print(f"Số lượng permalink_url: {raw_data_playlist['permalink_url'].nunique()}")
print(f"Số lượng uri: {raw_data_playlist['uri'].nunique()}")

Số lượng id: 1499
Số lượng permalink_url: 1499
Số lượng uri: 1499


Kiểm tra thuộc tính có quan hệ gần tương đồng nhau

In [11]:
print(f"Số lượng thuộc tính created_at: {raw_data_playlist['created_at'].nunique()}")
print(f"Số lượng thuộc tính release_date: {raw_data_playlist['release_date'].nunique()}")
print(f"Số lượng thuộc tính published_at: {raw_data_playlist['published_at'].nunique()}")    
print(f"Số lượng thuộc tính display_date: {raw_data_playlist['display_date'].nunique()}")     

Số lượng thuộc tính created_at: 1499
Số lượng thuộc tính release_date: 206
Số lượng thuộc tính published_at: 4
Số lượng thuộc tính display_date: 1499


In [12]:
raw_data_playlist[['created_at', 'release_date', 'published_at', 'display_date']]

created_at          release_date published_at  \
0     2008-04-29T22:14:25Z                   NaN          NaN   
1     2008-05-01T04:13:55Z  2005-06-29T00:00:00Z          NaN   
2     2008-05-01T17:31:06Z                   NaN          NaN   
3     2008-05-14T22:50:46Z                   NaN          NaN   
4     2008-05-15T11:47:21Z  2008-01-01T00:00:00Z          NaN   
...                    ...                   ...          ...   
1494  2009-04-30T16:49:33Z                   NaN          NaN   
1495  2009-04-30T17:28:35Z                   NaN          NaN   
1496  2009-04-30T18:20:20Z                   NaN          NaN   
1497  2009-04-30T19:46:59Z                   NaN          NaN   
1498  2009-04-30T20:11:48Z  2009-01-16T00:00:00Z          NaN   

              display_date  
0     2008-04-29T22:14:25Z  
1     2008-05-01T04:13:55Z  
2     2008-05-01T17:31:06Z  
3     2008-05-14T22:50:46Z  
4     2008-05-15T11:47:21Z  
...                    ...  
1494  2009-04-30T16:49:33Z  
1495  2009-04-30T17:28:35Z  
1496  2009-04-30T18:20:20Z  
1497  2009-04-30T19:46:59Z  
1498  2009-04-30T20:11:48Z  

[1499 rows x 4 columns]

Bốn thuộc tính `created_at`, `release_date`, `published_at`, `display_date` có chung nét tương đồng về thời gian liên quan đến playlist. Ta có thể lựa chọn 1 thuộc tính phù hợp để có thể thực hiện xử lý hay trực quan hóa dữ liệu sau này (do các playlist hiển thị trên SoundCloud mới có gần như đủ thông tin nên ta có thể chọn `display_date`)

Ta xem những thuộc tính còn lại

In [13]:
print(f"Số lượng artwork_url: {raw_data_playlist['artwork_url'].nunique()}\n")
# kind
print(f"Số lượng kind: {raw_data_playlist['kind'].nunique()}")
print(f"Giá trị của thuộc tính kind: {raw_data_playlist['kind'].unique()}\n")
# secret_token
print(f"Số lượng secret_token: {raw_data_playlist['secret_token'].nunique()}")
print(f"Giá trị của secret_token: {raw_data_playlist['secret_token'].unique()}\n")
# public
print(f"Số lượng public: {raw_data_playlist['public'].nunique()}")
print(f"Giá trị của public: {raw_data_playlist['public'].unique()}\n")      
# sharing
print(f"Số lượng sharing: {raw_data_playlist['sharing'].nunique()}") 
print(f"Giá trị của sharing: {raw_data_playlist['sharing'].unique()}\n")
# tag_list
print(f"Số lượng tag_list: {raw_data_playlist['tag_list'].nunique()}\n") 
print(f"Giá trị của tag_list: {raw_data_playlist['tag_list'].unique()}")  
# description
print(f"Số lượng description: {raw_data_playlist['description'].nunique()}")
print(f"Giá trị của tag_list: {raw_data_playlist['description'].unique()}")

Số lượng artwork_url: 576

Số lượng kind: 1
Giá trị của thuộc tính kind: ['playlist']

Số lượng secret_token: 0
Giá trị của secret_token: [nan]

Số lượng public: 1
Giá trị của public: [ True]

Số lượng sharing: 1
Giá trị của sharing: ['public']

Số lượng tag_list: 432

Giá trị của tag_list: [nan '"hip hop" abstract breakbeats' 'electro house techno'
 'paradise garage larry levan mancuso david new york'
 'funk soul disco afro nu disco edir re edit' 'vocal'
 'booty ghettotech detroit electro electrofunk juke'
 'booty electro ghettotech detroit turntablism scratching dj "ghetto house" juke'
 'booty electro ghettotech detroit' 'downbeat lounge breakbeat'
 'Breakbeat' 'ambient electronica electronic slow'
 'singer songwrtier nashville pop rock acoustic'
 'instrumental beats electronic experimental hip-hop'
 'trance house progressive techno deep melodic uplifting tech-trance'
 'guitar "space rock"' 'ambient drone dronescape synthesizer'
 'electronic low-fi techno minimal glitch acid rave'
 '

Xem số lượng theo từng tên hãng thu âm

In [14]:
raw_data_playlist['label_name'].value_counts()

TheDancingSkeleton       17
doubleshockrecordings     8
BATTLESNAKES              6
Proyekto Records          5
AVoxx                     4
                         ..
Suicide Pop! Records      1
Gigolo Records            1
Fazed Music               1
Sonarpilot Audio          1
DJAnders                  1
Name: label_name, Length: 303, dtype: int64

Xem số lượng theo đường dẫn API và URL

In [15]:
raw_data_playlist['permalink'].value_counts()

remixes                              2
free                                 2
dj-mixes                             2
embliss-presents-mind-over-matter    1
ghetto-blaster                       1
                                    ..
murdergrin-interludes                1
transition-ep                        1
extend-expand                        1
cape-unlike                          1
missions-sleepy-time-music           1
Name: permalink, Length: 1496, dtype: int64

Xem số lượng theo tiêu đề playlist

In [16]:
raw_data_playlist['title'].value_counts()

Remixes                                17
Mixes                                  13
DJ Mixes                               10
Originals                               3
Tracks                                  3
                                       ..
Stuff                                   1
FGM T2E                                 1
Altered Images                          1
Radioplastic - Hole in your Soul EP     1
Disco Re-edit                           1
Name: title, Length: 1422, dtype: int64

In [17]:
raw_data_playlist['purchase_title'].value_counts()

DOWNLOAD THE ALBUM     1
FREE DOWNLOAD          1
Buy Vinyl              1
Beatport               1
FLAC | DONATION        1
Subscribe on iTunes    1
Bandcamp               1
Juno Download          1
Mix archives           1
Mixcloud               1
Buy tracks             1
FREE DOWNLOAD HERE     1
Name: purchase_title, dtype: int64

Xem số lượng phụ thuộc vào đối tượng có thể nhúng

In [18]:
raw_data_playlist['embeddable_by'].value_counts()

all     1431
none      50
me        18
Name: embeddable_by, dtype: int64

Xem số lượng phụ thuộc `license`

In [19]:
raw_data_playlist['license'].value_counts()

all-rights-reserved    1446
cc-by-nc-sa              14
cc-by                    12
cc-by-nc                 10
cc-by-sa                  6
cc-by-nc-nd               5
no-rights-reserved        5
cc-by-nd                  1
Name: license, dtype: int64

## 3. Tập dữ liệu tracks

### Kiểm tra tập dữ liệu vừa thu thập

In [20]:
raw_data_track = pd.read_csv("Api_data/track.csv", sep=',')
raw_data_track.head()

artwork_url caption  commentable  \
0  https://i1.sndcdn.com/artworks-000000009541-24...     NaN         True   
1  https://i1.sndcdn.com/artworks-000260720873-yk...     NaN         True   
2                                                NaN     NaN         True   
3                                                NaN     NaN         True   
4  https://i1.sndcdn.com/artworks-000343027047-vs...     NaN         True   

   comment_count            created_at  \
0           43.0  2007-08-03T20:34:54Z   
1           53.0  2007-09-20T12:23:26Z   
2           81.0  2012-07-23T07:45:52Z   
3            2.0  2008-05-03T20:01:12Z   
4            3.0  2018-05-01T22:37:15Z   

                                         description  downloadable  \
0        Yet another old file from a long time ago…           True   
1  Messing around with live, might try and make s...          True   
2  Done at some point in 2004 for a car commercia...         False   
3                                                NaN         False   
4  dj set from høuse @ Holvi, Helsinki, Finland 2...          True   

   download_count  duration  full_duration  ... visuals  \
0           390.0     66456          66456  ...     NaN   
1          1602.0    207516         207516  ...     NaN   
2             0.0    165156         165156  ...     NaN   
3             0.0    254066         254066  ...     NaN   
4             0.0   6880852        6880852  ...     NaN   

                                  waveform_url          display_date  \
0  https://wave.sndcdn.com/PB0NHbxpC6Qn_m.json  2007-08-03T20:34:54Z   
1  https://wave.sndcdn.com/zT2WjrRTo236_m.json  2007-09-20T12:23:26Z   
2  https://wave.sndcdn.com/yKR1Zwr1TpUt_m.json  2012-07-23T07:45:52Z   
3  https://wave.sndcdn.com/XnvlAYJnre0c_m.json  2008-05-03T20:01:12Z   
4  https://wave.sndcdn.com/ikAxs0jbuMjt_m.json  2018-05-01T22:37:15Z   

                                               media  \
0  {'transcodings': [{'url': 'https://api-v2.soun...   
1  {'transcodings': [{'url': 'https://api-v2.soun...   
2  {'transcodings': [{'url': 'https://api-v2.soun...   
3  {'transcodings': [{'url': 'https://api-v2.soun...   
4  {'transcodings': [{'url': 'https://api-v2.soun...   

                                         station_urn  \
0      soundcloud:system-playlists:track-stations:43   
1     soundcloud:system-playlists:track-stations:274   
2  soundcloud:system-playlists:track-stations:537...   
3    soundcloud:system-playlists:track-stations:4795   
4  soundcloud:system-playlists:track-stations:438...   

          station_permalink  \
0         track-stations:43   
1        track-stations:274   
2   track-stations:53787294   
3       track-stations:4795   
4  track-stations:438021198   

                                 track_authorization monetization_model  \
0  eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJnZW8iO...     NOT_APPLICABLE   
1  eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJnZW8iO...     NOT_APPLICABLE   
2  eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJnZW8iO...     NOT_APPLICABLE   
3  eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJnZW8iO...     NOT_APPLICABLE   
4  eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJnZW8iO...     NOT_APPLICABLE   

   policy                                               user  
0   ALLOW  {'avatar_url': 'https://i1.sndcdn.com/avatars-...  
1   ALLOW  {'avatar_url': 'https://i1.sndcdn.com/avatars-...  
2   ALLOW  {'avatar_url': 'https://i1.sndcdn.com/avatars-...  
3   ALLOW  {'avatar_url': 'https://i1.sndcdn.com/avatars-...  
4   ALLOW  {'avatar_url': 'https://i1.sndcdn.com/avatars-...  

[5 rows x 48 columns]

In [21]:
raw_data_track.shape

(1499, 48)

Do tập dữ liệu tracks có nhiều thuộc tính và trùng lặp khá nhiều với tập users và playlists nên ta sẽ chỉ liệt kê các thuộc tính.

In [22]:
raw_data_track.columns

Index(['artwork_url', 'caption', 'commentable', 'comment_count', 'created_at',
       'description', 'downloadable', 'download_count', 'duration',
       'full_duration', 'embeddable_by', 'genre', 'has_downloads_left', 'id',
       'kind', 'label_name', 'last_modified', 'license', 'likes_count',
       'permalink', 'permalink_url', 'playback_count', 'public',
       'publisher_metadata', 'purchase_title', 'purchase_url', 'release_date',
       'reposts_count', 'secret_token', 'sharing', 'state', 'streamable',
       'tag_list', 'title', 'track_format', 'uri', 'urn', 'user_id', 'visuals',
       'waveform_url', 'display_date', 'media', 'station_urn',
       'station_permalink', 'track_authorization', 'monetization_model',
       'policy', 'user'],
      dtype='object')

Kiểu tra số lượng mẫu

In [23]:
print(f"Số lượng id: {raw_data_track['id'].nunique()}")
print(f"Số lượng permalink_url: {raw_data_track['permalink_url'].nunique()}")
print(f"Số lượng uri: {raw_data_track['uri'].nunique()}") 
print(f"Số lượng urn: {raw_data_track['urn'].nunique()}")    
print(f"Số lượng waveform_url: {raw_data_track['waveform_url'].nunique()}\n") 

Số lượng id: 1499
Số lượng permalink_url: 1499
Số lượng uri: 1499
Số lượng urn: 1499
Số lượng waveform_url: 1499



Kiểm tra số lượng mẫu ở các thuộc tính gần tương quan nhau.

Xét 3 thuộc tính `created_at`, `release_date`, `display_date`:

In [24]:
raw_data_track[['created_at', 'display_date', 'release_date']].head()

created_at          display_date          release_date
0  2007-08-03T20:34:54Z  2007-08-03T20:34:54Z  2002-01-01T00:00:00Z
1  2007-09-20T12:23:26Z  2007-09-20T12:23:26Z  2007-01-01T00:00:00Z
2  2012-07-23T07:45:52Z  2012-07-23T07:45:52Z                   NaN
3  2008-05-03T20:01:12Z  2008-05-03T20:01:12Z                   NaN
4  2018-05-01T22:37:15Z  2018-05-01T22:37:15Z                   NaN

Xét 2 thuộc tính `full_duration` và `duration`:

In [25]:
raw_data_track[['id','full_duration', 'duration']]

id  full_duration  duration
0            43          66456     66456
1           274         207516    207516
2      53787294         165156    165156
3          4795         254066    254066
4     438021198        6880852   6880852
...         ...            ...       ...
1494     103535        7242501   7242501
1495      70771        4253678   4253678
1496      61994        3991092   3991092
1497      81805        3884878   3884878
1498     106236        3321573   3321573

[1499 rows x 3 columns]

Ta chú ý đến những track đã được hiển thị và đồ dài thật sự của track đó nên ta có thể giữ lại thuộc tính `display_date` và `full_duration`

Xét số lượng các thuộc tính còn lại

In [26]:
print(f"Số lượng artwork_url: {raw_data_track['artwork_url'].nunique()}\n")
# commentable
print(f"Số lượng commentable: {raw_data_track['commentable'].nunique()}")
print(f"Giá trị của thuộc tính commentable: {raw_data_track['commentable'].unique()}")
print(f"Phân bố thuộc tính commentable:\n {raw_data_track['commentable'].value_counts()}\n")
# kind
print(f"Số lượng kind: {raw_data_track['kind'].nunique()}")
print(f"Giá trị của thuộc tính kind: {raw_data_track['kind'].unique()}\n")
# secret_token
print(f"Số lượng secret_token: {raw_data_track['secret_token'].nunique()}")
print(f"Giá trị của secret_token: {raw_data_track['secret_token'].unique()}\n")
# public
print(f"Số lượng public: {raw_data_track['public'].nunique()}")
print(f"Giá trị của public: {raw_data_track['public'].unique()}\n")
# sharing
print(f"Số lượng sharing: {raw_data_track['sharing'].nunique()}")
print(f"Giá trị của sharing: {raw_data_track['sharing'].unique()}\n")
# tag_list
print(f"Số lượng tag_list: {raw_data_track['tag_list'].nunique()}\n")
# description
print(f"Số lượng description: {raw_data_track['description'].nunique()}\n")
# caption
print(f"Số lượng caption: {raw_data_track['caption'].nunique()}\n")
# state
print(f"Số lượng state: {raw_data_track['state'].nunique()}")
print(f"Giá trị của state: {raw_data_track['state'].unique()}\n")
# streamable
print(f"Số lượng streamable: {raw_data_track['streamable'].nunique()}")
print(f"Giá trị của streamable: {raw_data_track['streamable'].unique()}\n")
print(f"Phân bố thuộc tính streamable:\n {raw_data_track['streamable'].value_counts()}\n")      
# publisher_metadata
print(f"Số lượng publisher_metadata: {raw_data_track['publisher_metadata'].nunique()}")
print(f"Giá trị của publisher_metadata: {raw_data_track['publisher_metadata'].unique()}\n")

Số lượng artwork_url: 881

Số lượng commentable: 2
Giá trị của thuộc tính commentable: [ True False]
Phân bố thuộc tính commentable:
 True     1496
False       3
Name: commentable, dtype: int64

Số lượng kind: 1
Giá trị của thuộc tính kind: ['track']

Số lượng secret_token: 0
Giá trị của secret_token: [nan]

Số lượng public: 1
Giá trị của public: [ True]

Số lượng sharing: 1
Giá trị của sharing: ['public']

Số lượng tag_list: 562

Số lượng description: 968

Số lượng caption: 3

Số lượng state: 1
Giá trị của state: ['finished']

Số lượng streamable: 2
Giá trị của streamable: [ True False]

Phân bố thuộc tính streamable:
 True     1481
False      18
Name: streamable, dtype: int64

Số lượng publisher_metadata: 232
Giá trị của publisher_metadata: [nan
 "{'id': 274, 'urn': 'soundcloud:tracks:274', 'artist': 'Alexander Ljung', 'contains_music': True, 'publisher': 'Alexander Ljung', 'isrc': 'QM8DG1704073', 'writer_composer': 'Alexander Ljung'}"
 "{'id': 53787294, 'urn': 'soundcloud:tracks:537

In [27]:
# downloadable
print(f"Số lượng downloadable: {raw_data_track['downloadable'].nunique()}")
print(f"Giá trị của downloadable: {raw_data_track['downloadable'].unique()}\n")
print(f"Phân bố thuộc tính downloadable:\n {raw_data_track['downloadable'].value_counts()}\n")
# download_count
print(f"Số lượng download_count: {raw_data_track['download_count'].nunique()}\n")
# has_downloads_left
print(f"Số lượng has_downloads_left: {raw_data_track['has_downloads_left'].nunique()}")
print(f"Giá trị của has_downloads_left: {raw_data_track['has_downloads_left'].unique()}\n")
print(f"Phân bố thuộc tính has_downloads_left:\n {raw_data_track['has_downloads_left'].value_counts()}\n")
# publisher_metadata
print(f"Số lượng thuộc tính publisher_metadata: {raw_data_track['publisher_metadata'].nunique()}")

Số lượng downloadable: 2
Giá trị của downloadable: [ True False]

Phân bố thuộc tính downloadable:
 False    837
True     662
Name: downloadable, dtype: int64

Số lượng download_count: 254

Số lượng has_downloads_left: 2
Giá trị của has_downloads_left: [ True False]

Phân bố thuộc tính has_downloads_left:
 False    922
True     577
Name: has_downloads_left, dtype: int64

Số lượng thuộc tính publisher_metadata: 232
